pkg/collector/process_test.go

```diff
@@ -0,0 +1,291 @@
+		cp.Stop()
+	}()
+	cp.Start()
+}
+
+func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
+	address := Address{"udp", "4735"}
+	cp, _ := InitCollectingProcess(address, 1024, 0)
+	go func() {
+		time.Sleep(time.Second)
+		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		if err != nil {
+			t.Errorf("UDP Address cannot be resolved.")
+		}
+		conn, err := net.DialUDP("udp", nil, resolveAddr)
+		if err != nil {
+			t.Errorf("UDP Collecting Process does not start correctly.")
+		}
+		defer conn.Close()
+		conn.Write(validTemplateRecord)
+	}()
+	go func() {
+		time.Sleep(time.Second)
+		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		if err != nil {
+			t.Errorf("UDP Address cannot be resolved.")
+		}
+		conn, err := net.DialUDP("udp", nil, resolveAddr)
+		if err != nil {
+			t.Errorf("UDP Collecting Process does not start correctly.")
+		}
+		defer conn.Close()
+		conn.Write(validTemplateRecord)
+		time.Sleep(time.Second)
+		assert.Equal(t, 2, len(cp.clients), "There should be two tcp clients.")
+	}()
+	go func() {
+		time.Sleep(6 * time.Second)
+		cp.Stop()
+	}()
+	cp.Start()
+}
+
+func TestCollectingProcess_DecodeTemplateRecord(t *testing.T) {
+	cp := collectingProcess{}
+	cp.templatesMap = make(map[uint32]map[uint16][]*templateField)
+	cp.templatesLock = sync.RWMutex{}
+	cp.address = Address{"tcp", "4736"}
+	message, err := cp.decodePacket(bytes.NewBuffer(validTemplateRecord))
+	if err != nil {
+		t.Fatalf("Got error in decoding template record: %v", err)
+	}
+	assert.Equal(t, uint16(10), message.Version, "Flow record version should be 10.")
+	assert.Equal(t, uint32(1), message.ObsDomainID, "Flow record obsDomainID should be 1.")
+	assert.NotNil(t, message.Record, "Template record should be stored in message flowset")
+	assert.NotNil(t, cp.templatesMap[message.ObsDomainID], "Template should be stored in template map")
+	// Invalid version
+	templateRecord := []byte{0, 9, 0, 40, 95, 40, 211, 236, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 24, 1, 0, 0, 3, 0, 8, 0, 4, 0, 12, 0, 4, 128, 105, 255, 255, 0, 0, 218, 21}
+	message, err = cp.decodePacket(bytes.NewBuffer(templateRecord))
+	assert.NotNil(t, err, "Error should be logged for invalid version")
+	// Malformed record
+	templateRecord = []byte{0, 10, 0, 40, 95, 40, 211, 236, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 24, 1, 0, 0, 3, 0, 8, 0, 4, 0, 12, 0, 4, 128, 105, 255, 255, 0, 0}
+	cp.templatesMap = make(map[uint32]map[uint16][]*templateField)
+	message, err = cp.decodePacket(bytes.NewBuffer(templateRecord))
+	assert.NotNil(t, err, "Error should be logged for malformed template record")
+	if _, exist := cp.templatesMap[uint32(1)]; exist {
+		t.Fatal("Template should not be stored for malformed template record")
+	}
+}
+
+func TestCollectingProcess_DecodeDataRecord(t *testing.T) {
+	cp := collectingProcess{}
+	cp.templatesMap = make(map[uint32]map[uint16][]*templateField)
+	cp.templatesLock = &sync.RWMutex{}
```

Comment:
will this work now?

Reply:
My mistake, should use `sync.RWMutex' here.